### PR TITLE
[Core] Azure CLI live test pipeline

### PR DESCRIPTION
**Pull request by qwordy**
**History Notes:**  
(Fill in the following template if multiple notes are needed, otherwise PR title will be used for history note.)

[Component Name 1] (BREAKING CHANGE:) (az command:) make some customer-facing change.  
[Component Name 2] (BREAKING CHANGE:) (az command:) make some customer-facing change.

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [ ] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [ ] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

### CODE REVIEW

**Comment by yonzhan** [comment]
live test pipeline

File: CLITest.yml [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+      inputs:
+        versionSpec: '3.8'
+        addToPath: true
+        architecture: 'x64'
+    - bash: |
+        cd ..
+        pwd
+        ls
+        git clone https://github.com/Azure/azure-cli.git
+        git clone https://github.com/Azure/azure-cli-extensions.git
```
Comment (yungezz) [inline CLITest.yml]: can we run extension tests in a seperate pipeline?

Comment (qwordy) [reply CLITest.yml]: OK

**Comment by yonzhan** [comment]
live test pipeline

File: scripts/live_test/graph_email.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+import json
+
+USER = False
+
+
+def main():
+    app = msal.ConfidentialClientApplication(
+        'b1023c2f-3625-4726-a173-09449d190793',
+        authority='https://login.microsoftonline.com/54826b22-38d6-4fb2-bad9-b7b93a3e9c5a',
+        client_credential='y5VL3nj0Yu.zYcQnra~ck-PF.6fr-t9p3u'
```
Comment (bim-msft) [inline scripts/live_test/graph_email.py:13]: How about using KeyVault to store these credentials?

Comment (qwordy) [reply scripts/live_test/graph_email.py:13]: Oh. I will change the credential immediately.
I am using Azure Pipeline's secret variable. KeyVault is also a good place to store credentials.

**Comment by yungezz** [comment]
hi @qwordy , I would suggest checkin the pipeline as first stable version. Then updating later. The PR commit history too huge now.

**Comment by qwordy** [comment]
> 
> 
> hi @qwordy , I would suggest checkin the pipeline as first stable version. Then updating later. The PR commit history too huge now.

OK

File: CLITest.yml [hunk 3]
```diff
@@ -0,0 +1,316 @@
+# Some content of this file is generated.
+
+name: CLI TEST RUN $(USER_TARGET) $(USER_LIVE) $(Date:yyyyMMdd)$(Rev:.r)
```
Comment (haroldrandom) [inline CLITest.yml:3]: As far as I understand, thoses files are not related to user usage. So, it shoud be placed inside at the root level of project.

File: CLITest.yml [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+        Target: extension
+    
+      eventgrid:
+        Target: eventgrid
+    
+  pool:
+    vmImage: 'Ubuntu-16.04'
+  steps:
+    - task: UsePythonVersion@0
+      condition: succeededOrFailed()
```
Comment (fengzhou-msft) [inline CLITest.yml:213]: Try [custom condition](https://docs.microsoft.com/en-us/azure/devops/pipelines/process/conditions?view=azure-devops&tabs=yaml#enable-a-custom-condition): `or(eq(variables['USER_TARGET'], '${{ Target }}'), eq(variables['USER_TARGET'], ''))`

Comment (qwordy) [reply CLITest.yml:213]: Good idea. I'll try

File: CLITest.yml (hunk 3 above)
Comment (qwordy) [reply CLITest.yml:3]: I'll move it to live test folder

Comment (qwordy) [reply CLITest.yml]: Done

File: scripts/live_test/CLITest.yml [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+        pip install pytest-html
+        pip install pytest-rerunfailures
+        azdev setup -c azure-cli
+        # az login --service-principal --username $(APP_ID) --password $(PASSWORD) --tenant $(TENANT_ID)
+        az login -u azureclitest@azuresdkteam.onmicrosoft.com -p $(AZURECLITESTPASSWORD)
+        az account set -s 0b1f6471-1bf0-4dda-aec3-cb9272f09590
+        # Sequential
+        azdev test $(Target) $(USER_LIVE) --mark serial --xml-path test_results.sequential.xml --no-exitfirst -a "--json-report --json-report-summary --json-report-file=$(Target).report.sequential.json --html=$(Target).report.sequential.html --self-contained-html --reruns 3 -s"
+        # Parallel
+        azdev test $(Target) $(USER_LIVE) --mark "not serial" --xml-path test_results.parallel.xml --no-exitfirst -a "-n $(USER_PARALLELISM) --json-report --json-report-summary --json-report-file=$(Target).report.parallel.json --html=$(Target).report.parallel.html --self-contained-html --reruns 3 -s"
```
Comment (fengzhou-msft) [inline scripts/live_test/CLITest.yml:186]: why called "not serial" instead of "parallel"?

File: scripts/live_test/CLITest.yml [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+        pwd
+        ls
+      condition: eq(variables.Match, '1')
```
Comment (fengzhou-msft) [inline scripts/live_test/CLITest.yml:189]: Is this line used?

File: scripts/live_test/CLITest.yml (hunk 5 above)
Comment (qwordy) [reply scripts/live_test/CLITest.yml:186]: "not serial" is not a plain string. "not" is an operator. If I use "parallel", I need to force users to add marks for all tests. Only a small portion of tests need to be marked as serial.

File: scripts/live_test/CLITest.yml (hunk 6 above)
Comment (qwordy) [reply scripts/live_test/CLITest.yml:189]: Yes, it is.

File: scripts/live_test/upload.py [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+    # Create container
+    date = datetime.datetime.now().strftime('%Y%m%d%H%M%S')
+    if USER_LIVE == '--live':
+        mode = 'live'
+    elif USER_LIVE == '':
+        mode = 'replay'
+    else:
+        mode = ''
+    container = date + mode
+    cmd = 'az storage container create -n {} --account-name clitestresultstac --account-key {} --public-access container'
```
Comment (haroldrandom) [inline scripts/live_test/upload.py:25]: How often will this script get run? And it creates a container every time?

File: scripts/live_test/CLITest.yml [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+      inputs:
+        buildType: 'current'
+        downloadType: 'specific'
+        downloadPath: '$(System.ArtifactsDirectory)'
+    - bash: |
+        echo $(System.ArtifactsDirectory)
+        ls $(System.ArtifactsDirectory)
+        pwd
+        commit_id=`git ls-remote https://github.com/Azure/azure-cli.git HEAD`
+        pip install sendgrid
```
Comment (haroldrandom) [inline scripts/live_test/CLITest.yml:242]: Using sendgrid to send emial obey our Microsoft's policy? I am not sure.

File: scripts/live_test/upload.py (hunk 7 above)
Comment (qwordy) [reply scripts/live_test/upload.py:25]: Each time the pipeline is triggered. Yes, it does.

File: scripts/live_test/CLITest.yml (hunk 8 above)
Comment (qwordy) [reply scripts/live_test/CLITest.yml:242]: Neither am I.
